### DISCUSSION TITLE
next/link navigation not updating page data

### DISCUSSION THREAD

**Discussion by alex-r89** [vercel/next.js#22512] [Help]
Hi,

I am using `next/link` within a custom navigation component. This navigation component generates a list of links based on mapping over a data object. This is as follows (i've simplified the code for readability):

```js
      <ul>
        <li>
          <Link href='/account'>
            <a>Account Home</a>
          </Link>
        </li>
        {photos.map((photo) => (
          <li key={photo.id}>
            <Link
              href={{
                pathname: '/account/photo/[id]',
                query: { id: photo.id },
              }}>
              <a>
                {photo.name}
              </a>
            </Link>
          </li>
        ))}
      </ul>
```

All of these links are working correctly, and I am able to navigate to one of these dynamic routes, eg `/account/photo/1`, via this nav or directly in the address bar.

The issue is however, is that If I am within one of these routes, for example `/account/photo/1`, when I click on another one of these links within the map, the url in the address bar changes, but the page does not update - nothing on the page changes. This means that I am on the route `/account/photo/2` but seeing data for the `/account/photo/1` route. Clicking on a link outside of this map, for example the first link (`Account Home`) works fine, but none of the other links cause the page data to update.

I was just wondering if anyone knows why this is?

**Comment by alex-r89** [comment]
I've managed to fix this, but I am unsure as to why this issue was occurring:

In the parent of the above component, I extract the  `photo.id`  from the path using `useRouter`, as follows: 
`  const { query } = router`

I then get the query.id to be the photo id.

The issue seems to be that this query value is not changing when a user navigated in the map in the code above, even though the URL was changing. My assumption was that useRouter updating would cause the component to rerender  - apparently it does not.

I fix this by placing a check inside a `useEffect` hook which then updates a state value if the `query.id` changes. This now fixes the issue and my page data is updating correctly.

If anyone is able to explain why useRouter was not causing my component to update, even though its value was, that would be great!

**Comment by kuongknight** [comment]
I confirmed this issue happened on version 10.0.7
I updated href but query from  useRouter did not changed!

**Comment by brunobraga95** [comment]
+1 same problem here

**Comment by samdoidge** [comment]
The [following solution ]( https://flaviocopes.com/nextjs-refresh-state-navigation/) helped me:

```
export default function App({ Component, pageProps }) {
  return <Component {...pageProps} />
}
```
changed to:
```
import { useRouter } from 'next/router'

export default function App({ Component, pageProps }) {
  const router = useRouter()

  return <Component {...pageProps} key={router.asPath} />
}
```

**Reply by BraydenTW** [reply]
Worked for me! Thanks!

**Reply by Kerim-Willem** [reply]
Very happy with your answer @samdoidge, this worked for me!

**Reply by nauvalazhar** [reply]
This solution worked for me, but why? 🤔

**Reply by bmarianome** [reply]
Why does that work?

**Reply by radikrisffnext** [reply]
Same issue, but this one fixed. Interesting.

**Reply by allenchuang** [reply]
> This solution worked for me, but why? 🤔

This solved it for me. I had the same issue- this is because of how react works as a SPA. It uses the virtualDOM to cache page changes so that the entire DOM doesn't do a full repaint, and it does so by caching each element on the DOM tree structure. So when you're re-using the same components on route pushes, it needs a unique key to recognize that they are different from each other, otherwise it just thinks the Component remains the same and just skips updating it.

This is the exact same reason you need to add 'key' when you are mapping over a list array.

**Reply by muhammadzaky15** [reply]
thanks you guys, this worked for me

**Reply by MastanAbdulkhaligli** [reply]
Thank U, this worked for me :)

**Reply by AngelMTR** [reply]
Worked, Thank You! <3

**Reply by danlesko** [reply]
+1 !

**Reply by ankha0109** [reply]
Thank you so much

**Comment by procrastagonist** [comment]
This solution works great in my case [#9992](https://github.com/vercel/next.js/issues/9992#issuecomment-615402511)
